test-upstream-merge/action.yml

```diff
@@ -0,0 +1,236 @@
+        fetch-depth: 0
+        token: ${{ inputs.gh-ci-token }}
+
+    - name: Get upstream repository name from URL
+      shell: bash
+      id: get-repository-name
+      run: |
+        echo "repository-name=$(echo '${{ inputs.upstream-repo-url }}' | sed 's|https://github.com/||g')" >> $GITHUB_OUTPUT
+
+    - name: Checkout upstream repository
+      uses: actions/checkout@v4
+      with:
+        repository: ${{ steps.get-repository-name.outputs.repository-name }}
+        ref: ${{ inputs.upstream-branch }}
+        path: upstream
+        fetch-depth: 1
+        token: ${{ inputs.gh-ci-token }}
+
+    - name: Get target last commit SHA
+      shell: bash
+      id: get-target-sha
+      run: |
+        cd target
+        git log -n 3
+        echo "sha=$(git rev-parse --short HEAD)" >> $GITHUB_OUTPUT
+
+    - name: Get upstream last commit SHA
+      shell: bash
+      id: get-upstream-sha
+      run: |
+        cd upstream
+        git log -n 3
+        echo "sha=$(git rev-parse --short HEAD)" >> $GITHUB_OUTPUT
+
+    - name: Check if merge is needed
+      shell: bash
+      id: check-if-merge-needed
+      run: |
+        cd target
+        set +e
+        merge_needed="no"
+        git merge-base --is-ancestor "${{ steps.get-upstream-sha.outputs.sha }}" HEAD
+        if [[ $? != 0 ]]; then
+          merge_needed="yes"
+        fi
+        set -e
+        echo "merge=${merge_needed}" >> $GITHUB_OUTPUT
+
+    - name: Remove existing upstream merge branch if nothing to do
+      shell: bash
+      run: |
+        if [[ "${{ steps.check-if-merge-needed.outputs.merge }}" != "yes" ]]; then
+          echo "Upstream SHA is an ancestor of target branch.  There is nothing to do.  Removing existing auto merge branch..."
+          set +e
+          git push origin :${pr_branch}
+          set -e
+        fi
+
+    - name: Add upstream
+      shell: bash
+      run: |
+        if [[ "${{ steps.check-if-merge-needed.outputs.merge }}" != "yes" ]]; then
+          echo "There is nothing to do."
+          exit 0
+        fi
+
+        cd target
+        git remote add upstream ${{ inputs.upstream-repo-url }}
+        git fetch upstream
+        git remote -v
+        git remote show upstream
+
+    - name: Merge upstream
+      shell: bash
+      id: create-pull-request
+      # yamllint enable rule:line-length
+      run: |
+        if [[ "${{ steps.check-if-merge-needed.outputs.merge }}" != "yes" ]]; then
+          echo "There is nothing to do."
+          exit 0
+        fi
+
+        cd target
+        git config user.name "${{ inputs.git-commit-author }}"
+        git config user.email "${{ inputs.git-commit-author }}"
+
+        # Name of PR branch
+        pr_branch="a0-ops-upstream-automerge"
+
+        # Get latest common ancestor
+        merge_base=$(git merge-base "${{ inputs.target-branch }}" "upstream/${{ inputs.upstream-branch }}")
+
+        # Remove existing branch
+        set +e
+        git push origin :${pr_branch}
```

Comment:
So this would cover scenario in which there's already pull request created? In such case, old one will be closed, and new one opened?

Reply:
Yes, exactly.  And here's the question if we should do that?  On one hand it might make some mess by creating so many PR's but on the other hand it'll make things transparent.
However, we might actually `git push --force` the branch and update description of existing PR as well.

Reply:
We need to have one such PR per repo only. Otherwise, there will be chaos which one team should work on. Closing current PR is not ideal, ideally, we should just do nothing this workflow-wise

Reply:
Yeah, so that's how it's done now.  Branch will be removed (and there's a `sleep 5` afterwards just in case) and PR will be closed as well.